eclipse-theia/theia
# rework build system

#### What it does

Fixes: https://github.com/eclipse-theia/theia/issues/7275.

Simplify `scripts/compile-references.js`. TypeScript Server now properly
understand TypeScript references for navigation, so we can simplify our
`tsconfig` setup.

Rewrite some `npm/yarn` scripts to do less on install (`yarn` or `yarn
install`). It will only run the required `prepare` scripts across our
packages, wire `tsconfig.json` files and build the TypeScript sources.

To build the example applications you need to run one of `yarn browser
build`, `yarn electron build` or `yarn build:examples`.

Linting is done as part of CI, but it won't automatically run locally
anymore. You can run `yarn lint` to lint your code. For the best
experience you should also enable the ESLint extension in your IDE.

#### How to test

Installation:

- `yarn` alone should not build the example applications anymore, nor should it lint.

Building:

- `yarn build` should build TypeScript and both apps.
- `yarn compile` should build TypeScript.
- `yarn browser build` should build both TypeScript and the browser app.
- `yarn electron build` should build both TypeScript and the electron app.
- `yarn build:examples` should build both apps.

Watching:

- `yarn watch` should watch TypeScript and both example apps.
- `yarn watch:ts` should watch only TypeScript.
- `yarn browser watch` should watch both TypeScript and the browser app.
- `yarn electron watch` should watch both TypeScript and the electron app.

Linting:

- `yarn lint` should lint TypeScript sources.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
